lib/listen/thread.rb

```diff
@@ -0,0 +1,45 @@
+require 'thread'
+
+require_relative 'logger'
+
+module Listen
+  module Thread
+    class << self
+      # Creates a new thread with the given name.
+      # Any exceptions raised by the thread will be logged with the thread name and complete backtrace.
+      def new(name)
+        thread_name = "listen-#{name}"
+
+        caller_stack = caller
+        ::Thread.new do
+          begin
+            yield
+          rescue Exception => ex
+            _log_exception(ex, thread_name, caller_stack)
```

Comment:
Maybe consider https://rubyapi.org/2.7/o/thread#method-i-report_on_exception

Reply:
Can you clarify your suggestion? Are you suggesting to not log ourselves and instead fall into the Ruby code? Or to check that boolean and do something different here?

Reply:
I just wonder what is the reason why we use our own thread wrapper, is it sufficient to use `report_on_exception`?

Reply:
Thanks for the clarification. I don't think the `report_on_exception` feature is sufficient. It writes directly to `$stderr`, which is better than nothing, but in general it won't appear in the `logger` with proper formatting. (In kubernetes will wind up swept into the `logger` at least--because they capture `$stderr` and `$stdout`--but in all cases, the log formatting will be missing. No date stamp, no severity. At our company we use JSON logging and of course this won't be formatted as valid JSON.)

I do like that Ruby added the `report_on_exception` feature, but I view it as a safety net that may help you see _something_ if you forgot to rescue and log at the top of the thread. Anecdote: over 10 years ago when we were running in production on Ruby 1.8.7, we had some terrible mystery bugs where "the process just stopping working". We eventually traced these down to exceptions getting raised in some worker threads. Those were intermittent, and getting silently ignored by Ruby, so it took much longer than it should have to figure out! From then on, we _always_ put `rescue`-and-log at the top of our threads and fibers.

Aside: I think the Ruby language should have the idea of a "default logger" for the thread/fiber/process and it should log there and *never* write to $stderr directly. Instead, the default for the "default logger" could do that, with degenerate formatting (don't show datestamp and severity, for example) so it looks about like it looks now when they call `$stderr.puts`. But applications that have their own logging strategy--which every production application does--could set Ruby default logger to point there so Ruby's messages would get plumbed to the logs with proper formatting.

Reply:
All makes sense.